NRLMMD-GEOIPS/geoips
# geoips#11 geoips - installation fails on mac

# Related Issue
NRLMMD-GEOIPS/geoips#11

# Reviewer Checklist

### Ensure you are logged into GitHub

### Confirm all requirements are met for pull request
* Ensure appropriate / sufficient content
    * **Title format**: geoips#<issue_num> <modified_repo_name> - \<Short description>
        * Example: geoips#8 geoips_template_plugin - Updating GEOIPS_REPO_URL to github
    * **Included Issue**: Link to related Issue included at the beginning of pull request
        * Example: GEOIPS/geoips#8
    * **CHANGELOG updated**: Functionality included in pull request is ALSO referenced in CHANGELOG.md
    * **Demonstrated Testing**: Proper testing / output was demonstrated for functionality updates
    * **Included Outputs**: Imagery is included in "Output" section for new or changed product outputs (for reports!)
* Ensure all appropriate tags / attributes added along right-hand side of pull request
    * **Label**: Issue ID Label, with color #00FF00 (bright green): geoips#<issue_num>
        * Example: geoips#8
    * **Projects**: GeoIPS - All Repos and All Functionality
        * Other projects allowed as appropriate
* Ensure Related Issue is finalized appropriately (follow link above)
    * **Check Issue Label**: Issue ID label added to Issue in bright green
    
### Once all items in checklist have been confirmed:
* **Approve pull request**
    * Click "Files changed" tab
    * Click green "Review changes" button
    * Select "Approve" option
    * Add message if desired
    * Click green "Submit review" button
    * Project status will automatically be updated in Project "GeoIPS2 - All Repos and All Functionality" when pull request is approved.
 
# Testing Instructions
- Remove your current `geoips_conda` environment `conda env remove --name geoips_conda`.
- Reinstall and test GeoIPS, allowing it to install conda for itself.
- Delete $GEOIPS_DEPENDENCIES_DIR
- Ensure that you have conda installed somewhere and that it is initialized (set up in your `.bash_profile`)
- Reinstall and test GeoIPS, but don't allow it to install conda

This should be tested on Windows, Mac, and Linux

# Summary
This PR fixes two issues.
- Broken installation process on Mac
- Difficulty installing with a system/user installation of conda

This updates `setup/geoips_conda_init_setup` to work well on all operating systems by removing reliance on sourcing `.bashrc`. Sourcing `.bashrc` is not possible on all systems from within a script. Instead, we use `eval "$(conda shell.bash activate geoips_conda)"`.

The script will default to trying to use conda as installed by GeoIPS if it is found, but will fall back to the first `conda` found in `$PATH` if it is found.

NRLMMD-GEOIPS/geoips#6,8,9,11 - Streamline installation process, support Mac installation

### Installation and Test
* **base_install_and_test.sh**
    * Exit immediately if GEOIPS_BASEDIR or GEOIPS_REPO_URL are not defined
    * Comment out several sections of installation, to reduce time and disk space
        * natural-earth-vector data download (will rely on latest shapefiles during cartopy processing)
        * natural-earth-vector linking to ~/.local/share/cartopy
            * Will NOT reinstate this step - cartopy supports CARTOPY_DATA_DIR as of 6 August 2021
        * vim8 installation (only for use of vim8 plugins to help with following style guides)
        * vim8 plugin installation
        * seviri setup
    * Remove BASECONDAPATH from conda cartopy installation (conda will be in PATH)
* **setup.sh**
    * To support Mac installations, use "uname -m" when determining filenames for
        rclone and miniconda3 installation
* **geoips_conda_init_setup**
    * To support Mac installations, use $(conda shell.bash activate geoips_conda) when activating
        conda vs "conda geoips_conda activate"
* **color_prompt**
    * Add "$CONDA_PROMPT_MODIFIER" to $PS1
* **repo_clone_update_install.sh**
    * If GEOIPS_TESTDATA_DIR, GEOIPS_PACKAGES_DIR, or GEOIPS_DEPENDENCIES_DIR are set, use those,
        otherwise default to placing under $GEOIPS_BASEDIR
* **README.md**
    * Update github.com GEOIPS_ACTIVE_BRANCH from dev to main


# Outputs
Linux, using existing conda:
```
Confirm environment variables point to desired installation parameters:
    GEOIPS_BASEDIR:       $HOME/geoproc2
    GEOIPS_CONFIG_FILE:   $HOME/geoproc2/geoips_packages/geoips/setup/config_geoips
    GEOIPS_ACTIVE_BRANCH: dev
    which conda (should point geoips_dependencies/bin): $HOME/geoproc/geoips_dependencies/miniconda3/condabin/conda
    which python (should point to miniconda3 envs/geoips_conda): $HOME/geoproc/geoips_dependencies/miniconda3/envs/geoips_conda/bin/python

Package: geoips_base
Total run time: 405 seconds
Number data types run: 3
Number data types failed: 0

full return: 0

# No miniconda3 in geoips_dependencies, only using existing conda:
bash$ ls $HOME/geoproc2/geoips_dependencies
bin  rclone  seviri_wavelet


```

Linux, reinstalling conda
```
# Remove geoips_conda so we don't accidentally use the old one:
$ conda env remove --name geoips_conda
Remove all packages in environment ~/geoproc/geoips_dependencies/miniconda3/envs/geoips_conda:

# Old conda install active:
bash$ which conda
~/geoproc/geoips_dependencies/miniconda3/bin/conda
# No geoips_conda env python:
$ which python
~/geoproc/geoips_dependencies/miniconda3/bin/python

# After re-installation in geoproc2:
Confirm environment variables point to desired installation parameters:
    GEOIPS_BASEDIR:       ~/geoproc2
    GEOIPS_CONFIG_FILE:   ~/geoproc2/geoips_packages/geoips/setup/config_geoips
    GEOIPS_ACTIVE_BRANCH: dev
    which conda (should point geoips_dependencies/bin): ~/geoproc2/geoips_dependencies/bin/conda
    which python (should point to miniconda3 envs/geoips_conda): ~/geoproc2/geoips_dependencies/miniconda3/envs/geoips_conda/bin/python

Thu Jul 28 09:46:08 PDT 2022  Running  abi.config_based_output.sh
        Return: 0
Thu Jul 28 09:50:14 PDT 2022  Running  abi.static.Visible.imagery_annotated.sh
        Return: 0
Thu Jul 28 09:50:35 PDT 2022  Running  test_interfaces
        Return: 0

Package: geoips_base
Total run time: 390 seconds
Number data types run: 3
Number data types failed: 0

full return: 0
```

Windows test with fresh install:
```
Thu Jul 28 09:05:47 PDT 2022  Running  abi.config_based_output.sh
        Return: 0
Thu Jul 28 09:09:09 PDT 2022  Running  abi.static.Visible.imagery_annotated.sh
        Return: 0
Thu Jul 28 09:09:20 PDT 2022  Running  test_interfaces
test_interfaces
        Return: 0

Package: geoips_base
Total run time: 261 seconds
Number data types run: 3
Number data types failed: 0

full return: 0
```


# Potential remaining problems
This will not fail if conda is not found at all. This is because the script is called both in cases where we would expect conda to be installed and in cases where we wouldn't. It might be good to do a bit more work to make the install process fail if conda is not available.
  
# Individual Commits
* [ Move auto-populated commits from beginning of pull request to this section, or just remove altogether. ]
